src/analyses/local_control_flow_history.h

```diff
@@ -0,0 +1,203 @@
+/*******************************************************************\
+
+Module: Abstract Interpretation
+
+Author: Martin Brain, martin.brain@cs.ox.ac.uk
+
+\*******************************************************************/
+
+/// \file
+/// Track function-local control flow for loop unwinding and path senstivity
+
+#ifndef CPROVER_ANALYSES_LOCAL_CONTROL_FLOW_HISTORY_H
+#define CPROVER_ANALYSES_LOCAL_CONTROL_FLOW_HISTORY_H
+
+#include "ai_history.h"
+
+/// Records all local control-flow decisions up to a limit of number of
+/// histories per location.  This gives a linear scaling between history limits
+/// and space and time used.  Note this does no interprocedural context so all
+/// calls will be merged to a single start point.
+
+/// The history is stored as a (shared) list of decisions, effectively a tree
+class local_control_flow_decisiont
+{
+public:
+  typedef ai_history_baset::locationt locationt;
+  locationt branch_location;
+  bool branch_taken;
+  typedef std::shared_ptr<const local_control_flow_decisiont> lcfd_ptrt;
+  lcfd_ptrt previous; // null at the start of the function
+
+  local_control_flow_decisiont(locationt loc, bool taken, lcfd_ptrt ptr)
+    : branch_location(loc), branch_taken(taken), previous(ptr)
+  {
+  }
+
+  bool operator<(const local_control_flow_decisiont &op) const;
+  bool operator==(const local_control_flow_decisiont &op) const;
+};
+
+/// Whether we track forwards and / or backwards jumps is compile-time fixed
+/// as it does not need to be variable and because there may be a *lot* of
+/// history objects created
+template <bool track_forward_jumps, bool track_backward_jumps>
+class local_control_flow_historyt : public ai_history_baset
+{
+protected:
+  typedef local_control_flow_decisiont::lcfd_ptrt lcfd_ptrt;
+
+  locationt current_loc;
+  lcfd_ptrt control_flow_decision_history;
+  size_t max_histories_per_location;
+
+  // A repeating idiom -- can cast from interface type to this type
+  // caller's responsibility to ensure correct types
+  std::shared_ptr<const local_control_flow_historyt<
+    track_forward_jumps,
+    track_backward_jumps>>
+  down_cast(trace_ptrt in) const
+  {
+    PRECONDITION(in != nullptr);
+    auto tmp = std::dynamic_pointer_cast<const local_control_flow_historyt<
+      track_forward_jumps,
+      track_backward_jumps>>(in);
+    PRECONDITION(tmp != nullptr);
```

Comment:
POSTCONDITION?

Reply:
I think it is a precondition of `down_cast` that this conversion works.  To quote the comment above `caller's responsibility to ensure correct types`.